conbench/entities/distribution.py

```diff
@@ -108,109 +111,160 @@ def get_distribution(benchmark_result, limit):
-                Distribution.context_id == first.context_id,
-            ]
-        )
+    For the start commit of the distribution, uses whatever commit_limit was used when
+    writing to the Distribution table.
+
+    Should not raise anything. If we can't find a z-score for some reason, the z_score
+    attribute will be None on that BenchmarkResult.
+    """
+    if not benchmark_results:
+        return
 
-    cols = [
-        Distribution.case_id,
-        Distribution.context_id,
-        Distribution.mean_mean,
-        Distribution.mean_sd,
-    ]
+    # Find the default branch first. If we can't find it, we'll just find the closest
+    # ancestor commit to each BenchmarkResult regardless of branch, which isn't ideal
+    # (see docstring) but it's good enough.
+    a_default_commit = benchmark_results[0].run.commit.get_fork_point_commit()
+    default_branch = a_default_commit.branch if a_default_commit else None
 
-    distributions = Session.query(*cols).filter(*where).all()
-    lookup = {f"{d.case_id}-{d.context_id}": d for d in distributions}
+    # Cache Distribution query results by commit/hardware/case/context, in case multiple
+    # input benchmark_results (e.g. from the same Run) share those attributes
+    cached_distributions = {}
 
     for benchmark_result in benchmark_results:
+        benchmark_result.z_score = None
         if benchmark_result.error:
             continue
 
-        d = lookup.get(f"{benchmark_result.case_id}-{benchmark_result.context_id}")
-        if d and d.mean_sd and benchmark_result.mean:
-            benchmark_result.z_score = (benchmark_result.mean - d.mean_mean) / d.mean_sd
+        closest_ancestor = get_closest_ancestor(benchmark_result, branch=default_branch)
+        if not closest_ancestor:
+            log.debug("Setting benchmark_result.z_score = None")
+            continue
+
+        commit = closest_ancestor.id
+        hardware = benchmark_result.run.hardware.hash
+        case = benchmark_result.case_id
+        context = benchmark_result.context_id
+
+        if (commit, hardware, case, context) in cached_distributions:
+            distribution = cached_distributions[(commit, hardware, case, context)]
+        else:
+            distribution = (
+                Session.query(Distribution)
+                .filter_by(
+                    commit_id=commit,
+                    hardware_hash=hardware,
+                    case_id=case,
+                    context_id=context,
+                )
+                .first()
+            )
+            cached_distributions[(commit, hardware, case, context)] = distribution
+
+        if (
+            benchmark_result.mean is not None
+            and distribution
+            and distribution.mean_mean is not None
+            and distribution.mean_sd  # is positive
+        ):
+            benchmark_result.z_score = (
+                benchmark_result.mean - distribution.mean_mean
+            ) / distribution.mean_sd
```

Comment:
1. thank you for `d` -> `distribution`; much more readable
2. I see this is the way it already was, so this is fine, but maybe we should talk about whether using `benchmark_result.mean` is the best choice. I see a few considerations:
  a. Median would be less influenced by outlier iterations. But [arguably] outliers matter here, so mean is more apt. Median plus separate calculations for spread would be more robust, if we want to go there.
  b. I think there's value here to using the z-scores of the raw iteration times instead of the means; the more means we take, the more the central limit theorem will mute interesting signal. The nice part is then handling spread becomes easier; we just return all the data instead of worrying about summary stats of summary stats of summary stats.

Not sure that's all right, but I do think this is an area worth more thought. If others agree (?), let's open an issue.

Reply:
When calculating `mean_mean`, couldn't we do a weighted-by-N mean of means in order to get the same result as taking the mean of the raw times? And if so, maybe we could do that with SD too (I'm rusty on stats)? The only trouble with using raw data is that Distribution is (I think) intended as a performance enhancer by calculating things quickly upon write, not read